infra/Pos/Network/CLI.hs

```diff
@@ -0,0 +1,196 @@
+  Defaults
+-------------------------------------------------------------------------------}
+
+-- | The topology we assume when no topology file is specified
+defaultTopology :: Y.Topology
+defaultTopology = Y.TopologyBehindNAT defaultDnsDomains
+
+-- | The default DNS domains used for relay discovery
+--
+defaultDnsDomains :: DnsDomains
+-- TODO: Give this a proper value
+defaultDnsDomains = DnsDomains [["todo.defaultDnsDomain.com"]]
+
+{-------------------------------------------------------------------------------
+  Interpreter
+-------------------------------------------------------------------------------}
+
+-- | Interpreter for the network config opts
+intNetworkConfigOpts :: NetworkConfigOpts -> IO T.NetworkConfig
+intNetworkConfigOpts cfg@NetworkConfigOpts{..} = do
+    parsedTopology <- case networkConfigOptsTopology of
+                        Nothing -> return defaultTopology
+                        Just fp -> parseTopology fp
+    ourTopology <- case parsedTopology of
+      Y.TopologyStatic allStaticallyKnownPeers ->
+        fromPovOf cfg allStaticallyKnownPeers networkConfigOptsSelf
+      Y.TopologyBehindNAT dnsDomains ->
+        return $ T.TopologyBehindNAT dnsDomains
+      Y.TopologyP2P ->
+        return $ T.TopologyP2P
+      Y.TopologyTransitional ->
+        return $ T.TopologyTransitional
+    return T.NetworkConfig {
+        ncTopology    = ourTopology
+      , ncDefaultPort = networkConfigOptsPort
+      , ncSelfName    = networkConfigOptsSelf
+      }
+
+-- | Perspective on 'AllStaticallyKnownPeers' from the point of view of
+-- a single node
+fromPovOf :: NetworkConfigOpts
+          -> Y.AllStaticallyKnownPeers
+          -> Maybe NodeName
+          -> IO T.Topology
+fromPovOf _ _ Nothing =
+    throwM NetworkConfigSelfUnknown
+fromPovOf cfg allPeers (Just self) = do
+    -- TODO: Do we want to allow to override the DNS config?
+    resolvSeed <- DNS.makeResolvSeed DNS.defaultResolvConf
+    DNS.withResolver resolvSeed $ \resolver -> do
+      selfMetadata <- metadataFor allPeers self
+      selfPeers    <- mkPeers resolver (Y.nmRoutes selfMetadata)
+      let selfType = Y.nmType selfMetadata
+      return $ T.TopologyStatic selfType (peersFromList selfPeers)
+  where
+    mkPeers :: DNS.Resolver -> Y.NodeRoutes -> IO [(T.NodeType, Alts NodeId)]
+    mkPeers resolver (Y.NodeRoutes routes) = mapM (mkAlts resolver) routes
+
+    mkAlts :: DNS.Resolver -> Alts NodeName -> IO (T.NodeType, Alts NodeId)
+    mkAlts _ [] = throwM $ EmptyListOfAltsFor self
+    mkAlts resolver alts@(firstAlt:_) = do
+        -- We silently assume all alternatives have the same type
+        altsType <- Y.nmType <$> metadataFor allPeers firstAlt
+        (altsType,) <$> mapM (resolveNodeName cfg resolver) alts
+
+-- | Resolve node name to IP address
+--
+-- We do this when reading the topology file so that we detect DNS problems
+-- early, and so that we are using canonical addresses (IP addresses) for
+-- node IDs.
```

Comment:
Note also that we should perhaps document that we don't ever want to use this `resolveNodeName` stuff in the behind NAT cases. It's important for edge nodes that we start up even if there are DNS problems that make the backend useless (since it's the GUI that reports user errors and it needs to be able to talk to the backend), whereas for core/relay nodes it's perfectly sensible to fail early and loudly.